### ISSUE TITLE
[FEATURE] enable actors to send async messages

### ISSUE THREAD

**Issue by stevelr** [wasmCloud/wasmcloud-otp#365]
It would be very useful for actors to be able to send an async message (equivalent to nats publish vs. current request/response) and continue execution. This would allow simplifying actor code that needs to send notifications and events, or fan-out a chain of events.

Addressing this touches several areas that will require design discussions, including syntax of invocation (codegen). Filing under wasmcloud_otp for now in part because a useful first step may be to adapt host_call to handle this case.

**Comment by stevelr** [comment]
I started to implement this on the code generator side, using a host function `__host_send` with the same signature as `__host_call`. The expectation from the client is that the response is `!= 1`, the __host_error_len and __host_error describe the error (same as for `__host_call`), and if the response is `== 1`, `__host_response_len` and `__host_response` are ignored, and `Ok(())` is returned to the caller.

**Comment by stevelr** [comment]
I thought of another way to implement async calls from actors that is a bit more generic, and wouldn't require generating all new interfaces. It would require putting the code you want to run into a separate function, and passing the function name and arg data to the host, which then invokes the function using wasmex/wasmer api, on a different otp thread.
The rust code  might look something like this...

```
// existing actor function handler that wants to invoke "my_async_function" asynchronously
async fn handler(ctx: &Context, arg: HandlerArg) {

    // Thing - the parameter to the async function, is any serializable object.
    // Parameters to async args must be self-contained aka "owned"
    let t: Thing = Thing::default();

    // invoke my_async_function in a separate OTP thread.
    // the call_async function serializes the arg, and passes the name of the async function
    // and the serialized arg to a host function, which queues up a call to 
    // "my_async_function_wrapper" (note addition of _wrapper suffix) on a new OTP thread
    call_async("my_async_function", t);

    // the actor function continues from here without waiting for call_async to return
    debug!("my_async_function may or may not have started..");
}


// this user-written function runs asynchronously, possibly in a separate thread from "handler"
// it is not called directly by the host. it is called by my_async_function_wrapper below.
#[async_export]
fn my_async_function(arg: Thing) {
    debug!("running inside async-function");
    // process Thing
}

// this function is generated by the '#[async_export]' macro
// (it's pretty simple and we could figure out an analogous approach for non-rust actor languages)
// it has an exported symbol because this is the function that the host invokes on the alternate thread.
export fn my_async_function_wrapper(ptr *u32, len: u32) {
    // todo: convert ptr and len to Vec<u8> (unsafe rust)
    // todo: deserialize Vec<u8> to Thing
    // then call the function
    my_async_function(arg);
}

```

**Comment by stevelr** [comment]
One advantage of this versus the idea we had of modifying code-generated interfaces to add async send functions is that this works even without invoking a host_send message- it can work for an actor's compute-intensive operations that don't involve calling another entity. For example, ray-tracing actor could divide up a scene into 16 different sub-regions, and invoke the async function 16 times, then wait on a subscription to gather results - effectively creating a map-join operation. A game-playing actor (chess, poker, whatever) could receive a copy of the game state and call an async function to compute the next suggested move and then send the calculated move to a subscriber.  The actor could return to its caller after invoking the async function, almost immediately, because it doesn't need to wait for the return of the async function.

**Comment by stevelr** [comment]
The second idea above (exported async function) is similar to a special case of actor-invoking-actor, where the invoked actor is always "self", and the invocation isn't signed and always runs in the same host, (whereas for `__host_send`, the Invocation would use nats queue subscription which may use alternate hosts). To achieve parallelism, there would need to be multiple instances of the actor on the same host, and when the host calls the exported async function, it would choose from the available set of instances.

If the host only has one instance of the actor, execution of the async function would be deferred until after the sync rpc returns and the actor is able to be scheduled to run again.

**Comment by autodidaddict** [comment]
All of this is really good stuff to consider, but the topic of "asynchronous actors" is actually far wider than simply allowing actors to send "fire-and-forget" messages. We don't need to achieve parallelism in actors anytime soon.

I think for now, the easiest thing to get running would be for the host to support the `__host_send` function and then for the Rust SDK to have `send` versions of all the functions an actor can call.
